equinox/nn/attention.py

```diff
@@ -0,0 +1,237 @@
+        """
+        super().__init__(**kwargs)
+        qkey, kkey, vkey, okey = jrandom.split(key, 4)
+
+        if key_size is None:
+            key_size = query_size
+        if value_size is None:
+            value_size = query_size
+        if qk_size is None:
+            qk_size = query_size // num_heads
+        if vo_size is None:
+            vo_size = query_size // num_heads
+        if output_size is None:
+            output_size = query_size
+
+        self.query_proj = Linear(
+            query_size, num_heads * qk_size, use_bias=use_query_bias, key=qkey
+        )
+        self.key_proj = Linear(
+            key_size, num_heads * qk_size, use_bias=use_key_bias, key=kkey
+        )
+        self.value_proj = Linear(
+            value_size, num_heads * vo_size, use_bias=use_value_bias, key=vkey
+        )
+        self.output_proj = Linear(
+            num_heads * vo_size, output_size, use_bias=use_output_bias, key=okey
+        )
+        self.dropout = Dropout(dropout_p)
+
+        self.num_heads = num_heads
+        self.query_size = query_size
+        self.key_size = key_size
+        self.value_size = value_size
+        self.output_size = output_size
+        self.qk_size = qk_size
+        self.vo_size = vo_size
+        self.use_query_bias = use_query_bias
+        self.use_key_bias = use_key_bias
+        self.use_value_bias = use_value_bias
+        self.use_output_bias = use_output_bias
+
+    def __call__(
+        self,
+        query: Array["seq_length", "query_size"],  # noqa: F821
+        key_: Array["seq_length", "key_size"],  # noqa: F821
+        value: Array["seq_length", "value_size"],  # noqa: F821
+        mask: Optional[
+            Array["num_heads", "seq_length", "seq_length"]  # noqa: F821
+        ] = None,
+        *,
+        key: Optional["jax.random.PRNGKey"] = None,
+        deterministic: Optional[bool] = None,
+    ) -> Array["seq_length", "output_size"]:  # noqa: F821
+        """**Arguments:**
+
+        - `query`: Query embedding. Should be a JAX array of shape
+            `(seq_length, query_size)`.
+        - `key_`: Key embedding. Should be a JAX array of shape
+            `(seq_length, key_size)`.
+        - `value`: Value embedding. Should be a JAX array of shape
+            `(seq_length, value_size)`.
+        - `mask`: Optional mask preventing attention to certain positions. Should be a
+            JAX array of shape `(num_heads, seq_length, seq_length)`.
+        - `key`: A `jax.random.PRNGKey` used for dropout. Unused if `dropout = 0`.
+            (Keyword only argument.)
+        - `deterministic`: As [`equinox.nn.Dropout.__call__`][]. (Keyword only
+            argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(seq_length, output_size)`.
+        """
+
+        seq_length, _ = query.shape
+        seq_length2, _ = key_.shape
+        seq_length3, _ = value.shape
+        if seq_length != seq_length2 or seq_length != seq_length3:
```

Comment:
sorry, one last thing - so with this condition (and the message below for masking), you will only be allowing for self-attention, and not cross attention

to allow for cross attention (allow for full encoder - decoder), you will need to change this to `seq_length2 != seq_length3`